### PR TITLE
ANTs - new ebuild 

### PR DESCRIPTION
**Pull request by TheChymera**
Ants is a rather cool and advanced neuroimaging normalization toolkit. Sadly, it comes without an install target, and upstream seems unwilling [(scroll to bottom of page)](https://brianavants.wordpress.com/2012/04/13/updated-ants-compile-instructions-april-12-2012/#comment-399) to provide one. 

Currently the ebuild fails with this [error message](https://bpaste.net/show/b7095b966ff3). And neurodebian seems to deal with the issue [like so](http://anonscm.debian.org/cgit/pkg-exppsy/ants.git/tree/debian/rules).

In the aforementioned upstream link, one of the authors seemed to imply that the absence of the install target can easily be accounted for by “make” ? any ideas what that is about?

### CODE REVIEW

**Comment by TheChymera** [comment]
I am reading up on CMake install targets, and it seems [(see step 3)](http://www.cmake.org/cmake-tutorial/) the only thing needed would be to patch some lines of `CMakeLists.txt`?

That file currently looks like so:

```
cmake_minimum_required(VERSION 2.8.5)
cmake_policy(VERSION 2.8.5)

set(LIBRARY_SOVERSION_INFO 2)
set(LIBRARY_VERSION_INFO 2.1.0)

set(LOCAL_PROJECT_NAME ANTS)
INCLUDE(InstallRequiredSystemLibraries)
## NOTE THERE SHOULD BE NO PROJECT STATEMENT HERE!
## This file acts as a simple switch to initiate
## two completely independant CMake build environments.

#-----------------------------------------------------------------------------
# Superbuild Option - Enabled by default
#                   Phase I:  ${LOCAL_PROJECT_NAME}_SUPERBUILD is set to ON, and the
#                             supporting packages defined in "SuperBuild.cmake"
#                             are built.  The last package in "SuperBuild.cmake"
#                             to be built is a recursive call to this
#                             file with ${LOCAL_PROJECT_NAME}_SUPERBUILD explicitly
#                             set to "OFF" to initiate Phase II
#
#                   Phase II: Build the ${LOCAL_PROJECT_NAME}, referencing the support
#                             packages built in Phase I.
#-----------------------------------------------------------------------------
option(${LOCAL_PROJECT_NAME}_SUPERBUILD "Build ${LOCAL_PROJECT_NAME} and the projects it depends on via SuperBuild.cmake." ON)
mark_as_advanced(${LOCAL_PROJECT_NAME}_SUPERBUILD)

option(${LOCAL_PROJECT_NAME}_USE_QT "Find and use Qt with VTK to build GUI Tools" OFF)
mark_as_advanced(${LOCAL_PROJECT_NAME}_USE_QT)

option(${LOCAL_PROJECT_NAME}_INSTALL_DEVELOPMENT "Install development support include and libraries for external packages." OFF)
mark_as_advanced(${LOCAL_PROJECT_NAME}_INSTALL_DEVELOPMENT)

option(COPY_SCRIPT_FILES_TO_BIN_DIR "Copy the script files to the ANTS bin directory." OFF)
mark_as_advanced(COPY_SCRIPT_FILES_TO_BIN_DIR)

set(CPACK_PACKAGE_NAME "ANTs")
set(CPACK_PACKAGE_VENDOR "CMake.org")
set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "ANTs - Advanced Normalization Tools")
set(CPACK_PACKAGE_VERSION 2.1.0)
set(CPACK_PACKAGE_VERSION_MAJOR 2)
set(CPACK_PACKAGE_VERSION_MINOR 1)
set(CPACK_PACKAGE_VERSION_PATCH 0)
set(CPACK_PACKAGE_INSTALL_DIRECTORY "ANTS_Install")
set(CPACK_BINARY_GENERATORS "DragNDrop TGZ TZ")
set(CPACK_PACKAGE_DESCRIPTION_FILE "${CMAKE_CURRENT_SOURCE_DIR}/README.txt")
set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "ANTs - robust image registration, segmentation and more")
set(CPACK_RESOURCE_FILE_LICENSE  "${CMAKE_CURRENT_SOURCE_DIR}/ANTSCopyright.txt")
set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README.txt")
#-----------------------------------------------------------------------------
# Superbuild script
#-----------------------------------------------------------------------------
if(${LOCAL_PROJECT_NAME}_SUPERBUILD)
  project(SuperBuild_${LOCAL_PROJECT_NAME})
  include("${CMAKE_CURRENT_SOURCE_DIR}/SuperBuild.cmake")
  include(CPack)   # This must always be last!
  return()
else()
  project(${LOCAL_PROJECT_NAME})
  include("${CMAKE_CURRENT_SOURCE_DIR}/${LOCAL_PROJECT_NAME}.cmake")
  include(CPack)   # This must always be last!
  return()
endif()

message(FATAL_ERROR "You should never reach this point !")

MESSAGE(STATUS "<<< Gentoo configuration >>>
Build type      ${CMAKE_BUILD_TYPE}
Install path    ${CMAKE_INSTALL_PREFIX}
Compiler flags:
C               ${CMAKE_C_FLAGS}
C++             ${CMAKE_CXX_FLAGS}
Linker flags:
Executable      ${CMAKE_EXE_LINKER_FLAGS}
Module          ${CMAKE_MODULE_LINKER_FLAGS}
Shared          ${CMAKE_SHARED_LINKER_FLAGS}\n")
```

Where should I put the `install (TARGETS whatever DESTINATION bin)` lines in there?

@kiwifb @jlec can you help me out maybe?

**Comment by jlec** [comment]
You need to place the install target lines next to the build target ones.

File: sci-biology/ants/ants-9999.ebuild [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+SLOT="0"
+LICENSE="BSD"
+KEYWORDS=""
+
+DEPEND="sci-libs/itk"
+RDEPEND="${DEPEND}"
+
+src_install() {
+	pwd
+	ls
```
Comment (jlec) [inline sci-biology/ants/ants-9999.ebuild]: Not needed.

File: sci-biology/ants/ants-9999.ebuild [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	cd "${P}_build/ANTS-build"
+	emake DESTDIR="${D}" install
```
Comment (jlec) [inline sci-biology/ants/ants-9999.ebuild]: That's the default.

Comment (TheChymera) [inline sci-biology/ants/ants-9999.ebuild]: that is strange, navigating to that dir after the build fails and running make install gives me this:

```
-- Installing: /usr/lib/libl_antsAffineInitializer.a
-- Installing: /usr/bin/antsAffineInitializer
-- Installing: /usr/lib/libl_SurfaceBasedSmoothing.a
-- Installing: /usr/bin/SurfaceBasedSmoothing
-- Installing: /usr/lib/libl_CreateJacobianDeterminantImage.a
-- Installing: /usr/bin/CreateJacobianDeterminantImage
-- Installing: /usr/lib/libl_ThresholdImage.a
-- Installing: /usr/bin/ThresholdImage
-- Installing: /usr/lib/libl_ResampleImage.a
-- Installing: /usr/bin/ResampleImage
-- Installing: /usr/lib/libl_sccan.a
-- Installing: /usr/bin/sccan
-- Installing: /usr/lib/libl_N4BiasFieldCorrection.a
-- Installing: /usr/bin/N4BiasFieldCorrection
-- Installing: /usr/lib/libl_N3BiasFieldCorrection.a
-- Installing: /usr/bin/N3BiasFieldCorrection
-- Installing: /usr/lib/libl_KellyKapowski.a
-- Installing: /usr/bin/KellyKapowski
-- Installing: /usr/lib/libl_antsRegistration.a
-- Installing: /usr/bin/antsRegistration
-- Installing: /usr/lib/libl_antsMotionCorrStats.a
-- Installing: /usr/bin/antsMotionCorrStats
-- Installing: /usr/lib/libl_antsMotionCorr.a
-- Installing: /usr/bin/antsMotionCorr
-- Installing: /usr/lib/libl_antsApplyTransforms.a
-- Installing: /usr/bin/antsApplyTransforms
-- Installing: /usr/lib/libl_LabelGeometryMeasures.a
-- Installing: /usr/bin/LabelGeometryMeasures
-- Installing: /usr/lib/libl_LabelClustersUniquely.a
-- Installing: /usr/bin/LabelClustersUniquely
-- Installing: /usr/lib/libl_Atropos.a
-- Installing: /usr/bin/Atropos
-- Installing: /usr/lib/libl_AverageImages.a
-- Installing: /usr/bin/AverageImages
-- Installing: /usr/lib/libl_StackSlices.a
-- Installing: /usr/bin/StackSlices
-- Installing: /usr/lib/libl_AverageAffineTransform.a
-- Installing: /usr/bin/AverageAffineTransform
-- Installing: /usr/lib/libl_AverageAffineTransformNoRigid.a
-- Installing: /usr/bin/AverageAffineTransformNoRigid
-- Installing: /usr/lib/libl_antsApplyTransformsToPoints.a
-- Installing: /usr/bin/antsApplyTransformsToPoints
-- Installing: /usr/lib/libl_antsAI.a
-- Installing: /usr/bin/antsAI
-- Installing: /usr/lib/libl_antsJointFusion.a
-- Installing: /usr/bin/antsJointFusion
-- Installing: /usr/lib/libl_ImageMath.a
-- Installing: /usr/bin/ImageMath
-- Installing: /usr/lib/libl_iMath.a
-- Installing: /usr/bin/iMath
-- Installing: /usr/lib/libl_ANTS.a
-- Installing: /usr/bin/ANTS
-- Installing: /usr/lib/libl_ANTSJacobian.a
-- Installing: /usr/bin/ANTSJacobian
-- Installing: /usr/lib/libl_PrintHeader.a
-- Installing: /usr/bin/PrintHeader
-- Installing: /usr/lib/libl_ResetDirection.a
-- Installing: /usr/bin/ResetDirection
-- Installing: /usr/lib/libl_ANTSUseLandmarkImagesToGetAffineTransform.a
-- Installing: /usr/bin/ANTSUseLandmarkImagesToGetAffineTransform
-- Installing: /usr/lib/libl_ANTSUseLandmarkImagesToGetBSplineDisplacementField.a
-- Installing: /usr/bin/ANTSUseLandmarkImagesToGetBSplineDisplacementField
-- Installing: /usr/lib/libl_ANTSUseDeformationFieldToGetAffineTransform.a
-- Installing: /usr/bin/ANTSUseDeformationFieldToGetAffineTransform
-- Installing: /usr/lib/libl_LaplacianThickness.a
-- Installing: /usr/bin/LaplacianThickness
-- Installing: /usr/lib/libl_SetOrigin.a
-- Installing: /usr/bin/SetOrigin
-- Installing: /usr/lib/libl_SetSpacing.a
-- Installing: /usr/bin/SetSpacing
-- Installing: /usr/lib/libl_SetDirectionByMatrix.a
-- Installing: /usr/bin/SetDirectionByMatrix
-- Installing: /usr/lib/libl_SurfaceCurvature.a
-- Installing: /usr/bin/SurfaceCurvature
-- Installing: /usr/lib/libl_ConvertScalarImageToRGB.a
-- Installing: /usr/bin/ConvertScalarImageToRGB
-- Installing: /usr/lib/libl_CreateWarpedGridImage.a
-- Installing: /usr/bin/CreateWarpedGridImage
-- Installing: /usr/lib/libl_MeasureImageSimilarity.a
-- Installing: /usr/bin/MeasureImageSimilarity
-- Installing: /usr/lib/libl_ConvertToJpg.a
-- Installing: /usr/bin/ConvertToJpg
-- Installing: /usr/lib/libl_ConvertImage.a
-- Installing: /usr/bin/ConvertImage
-- Installing: /usr/lib/libl_ConvertImagePixelType.a
-- Installing: /usr/bin/ConvertImagePixelType
-- Installing: /usr/lib/libl_ConvertInputImagePixelTypeToFloat.a
-- Installing: /usr/bin/ConvertInputImagePixelTypeToFloat
-- Installing: /usr/lib/libl_FitBSplineToPoints.a
-- Installing: /usr/bin/FitBSplineToPoints
-- Installing: /usr/lib/libl_AverageTensorImages.a
-- Installing: /usr/bin/AverageTensorImages
-- Installing: /usr/lib/libl_ImageSetStatistics.a
-- Installing: /usr/bin/ImageSetStatistics
-- Installing: /usr/lib/libl_MultiplyImages.a
-- Installing: /usr/bin/MultiplyImages
-- Installing: /usr/lib/libl_SmoothImage.a
-- Installing: /usr/bin/SmoothImage
-- Installing: /usr/lib/libl_ClusterImageStatistics.a
-- Installing: /usr/bin/ClusterImageStatistics
-- Installing: /usr/lib/libl_LabelOverlapMeasures.a
-- Installing: /usr/bin/LabelOverlapMeasures
-- Installing: /usr/lib/libl_LesionFilling.a
-- Installing: /usr/bin/LesionFilling
-- Installing: /usr/lib/libl_MeasureMinMaxMean.a
-- Installing: /usr/bin/MeasureMinMaxMean
-- Installing: /usr/lib/libl_WarpImageMultiTransform.a
-- Installing: /usr/bin/WarpImageMultiTransform
-- Installing: /usr/lib/libl_ComposeMultiTransform.a
-- Installing: /usr/bin/ComposeMultiTransform
-- Installing: /usr/lib/libl_MemoryTest.a
-- Installing: /usr/bin/MemoryTest
-- Installing: /usr/lib/libl_PermuteFlipImageOrientationAxes.a
-- Installing: /usr/bin/PermuteFlipImageOrientationAxes
-- Installing: /usr/lib/libl_ImageCompare.a
-- Installing: /usr/bin/ImageCompare
-- Installing: /usr/lib/libl_ResampleImageBySpacing.a
-- Installing: /usr/bin/ResampleImageBySpacing
-- Installing: /usr/lib/libl_CopyImageHeaderInformation.a
-- Installing: /usr/bin/CopyImageHeaderInformation
-- Installing: /usr/lib/libl_WarpTimeSeriesImageMultiTransform.a
-- Installing: /usr/bin/WarpTimeSeriesImageMultiTransform
-- Installing: /usr/lib/libl_ExtractSliceFromImage.a
-- Installing: /usr/bin/ExtractSliceFromImage
-- Installing: /usr/lib/libl_ExtractRegionFromImage.a
-- Installing: /usr/bin/ExtractRegionFromImage
-- Installing: /usr/lib/libl_ExtractRegionFromImageByMask.a
-- Installing: /usr/bin/ExtractRegionFromImageByMask
-- Installing: /usr/lib/libl_PasteImageIntoImage.a
-- Installing: /usr/bin/PasteImageIntoImage
-- Installing: /usr/lib/libl_TileImages.a
-- Installing: /usr/bin/TileImages
-- Installing: /usr/lib/libl_CreateTiledMosaic.a
-- Installing: /usr/bin/CreateTiledMosaic
-- Installing: /usr/lib/libl_CreateImage.a
-- Installing: /usr/bin/CreateImage
-- Installing: /usr/lib/libl_DenoiseImage.a
-- Installing: /usr/bin/DenoiseImage
-- Installing: /usr/lib/libl_WarpTensorImageMultiTransform.a
-- Installing: /usr/bin/WarpTensorImageMultiTransform
-- Installing: /usr/lib/libl_ReorientTensorImage.a
-- Installing: /usr/bin/ReorientTensorImage
-- Installing: /usr/lib/libl_RebaseTensorImage.a
-- Installing: /usr/bin/RebaseTensorImage
-- Installing: /usr/lib/libl_KellySlater.a
-- Installing: /usr/bin/KellySlater
-- Installing: /usr/lib/libl_CreateDTICohort.a
-- Installing: /usr/bin/CreateDTICohort
-- Installing: /usr/lib/libl_antsAlignOrigin.a
-- Installing: /usr/bin/antsAlignOrigin
-- Installing: /usr/lib/libl_simpleSynRegistration.a
-- Installing: /usr/bin/simpleSynRegistration
-- Installing: /usr/lib/libl_CompositeTransformUtil.a
-- Installing: /usr/bin/CompositeTransformUtil
-- Installing: /usr/lib/libl_antsMotionCorrDiffusionDirection.a
-- Installing: /usr/bin/antsMotionCorrDiffusionDirection
-- Installing: /usr/lib/libl_antsSliceRegularizedRegistration.a
-- Installing: /usr/bin/antsSliceRegularizedRegistration
-- Installing: /usr/lib/libl_CreateDisplacementField.a
-- Installing: /usr/bin/CreateDisplacementField
-- Installing: /usr/lib/libl_ANTSIntegrateVectorField.a
-- Installing: /usr/bin/ANTSIntegrateVectorField
-- Installing: /usr/lib/libl_ANTSIntegrateVelocityField.a
-- Installing: /usr/bin/ANTSIntegrateVelocityField
-- Installing: /usr/lib/libl_ConvertTransformFile.a
-- Installing: /usr/bin/ConvertTransformFile
-- Installing: /usr/lib/libl_compareTwoTransforms.a
-- Installing: /usr/bin/compareTwoTransforms
-- Installing: /usr/lib/libl_TimeSCCAN.a
-- Installing: /usr/bin/TimeSCCAN
-- Installing: /usr/lib/libl_TextureCooccurrenceFeatures.a
-- Installing: /usr/bin/TextureCooccurrenceFeatures
-- Installing: /usr/lib/libl_TextureRunLengthFeatures.a
-- Installing: /usr/bin/TextureRunLengthFeatures
-- Installing: /usr/lib/libl_ImageIntensityStatistics.a
-- Installing: /usr/bin/ImageIntensityStatistics
-- Installing: /usr/lib/libl_GetConnectedComponentsFeatureImages.a
-- Installing: /usr/bin/GetConnectedComponentsFeatureImages
-- Installing: /usr/lib/libl_DeNrrd.a
-- Installing: /usr/bin/DeNrrd
-- Installing: /usr/lib/libl_antsTransformInfo.a
-- Installing: /usr/bin/antsTransformInfo
-- Installing: /usr/lib/libl_antsUtilitiesTesting.a
-- Installing: /usr/bin/antsUtilitiesTesting
-- Installing: /usr/bin/jointfusion
-- Installing: /usr/bin/sa
-- Installing: /usr/bin/bl
```

And indeed I can see those files were placed there. How can I make this work via portage? also, how could I remove the files which I now installed? a bit many to rm y hand :-/

**Comment by TheChymera** [comment]
Wow, this actually works now, and doesn't bundle anything. Unexpectedly nice really.

What do you think @jlec, ready for merge?

**Comment by operatornormal** [comment]
Hmm, I got this e-build installing something with this kind of version
```
# Copyright 1999-2015 Gentoo Foundation
# Distributed under the terms of the GNU General Public License v2
# $Header: $

EAPI=5

inherit cmake-utils git-r3

DESCRIPTION="Advanced Normalitazion Tools for neuroimaging"
HOMEPAGE="http://stnava.github.io/ANTs/"
SRC_URI=""
EGIT_REPO_URI="git://github.com/stnava/ANTs.git"

SLOT="0"
LICENSE="BSD"
KEYWORDS=""

DEPEND="sci-libs/itk"
RDEPEND="${DEPEND}"

src_install() {
        # idea of this function and the sed regexp stolen from debian 
        # rules file, originally by Yaroslav Halchenko. 
        cd "../${P}_build" || die "build dir not found"
        dodir /usr/lib/ants
        install -t "${D}"usr/lib/ants bin/*
        dodir /usr/bin
        dosym "${D}"usr/lib/ants/ANTS /usr/bin/ANTS
        dosym "${D}"usr/lib/ants/WarpImageMultiTransform /usr/bin/WarpImageMulti
Transform
        cd ../"${P}"/Scripts || die "scripts dir not found"
        install -t "${D}"usr/lib/ants *
        cd "${D}"usr/lib/ants
        sed -ie 's,\([^"=]*\(/bin/ants\|ANTS/release/bin\)\),/usr/lib/ants,g' *.
sh *.pl && rm *.she *.ple
}
```

but then I do not know ants so I can't really verify if ANTs is functional or not because I don't have a faintest clue about how it should be used.

**Comment by TheChymera** [comment]
@operatornormal I asked upstream, and [it seems](https://github.com/stnava/ANTs/issues/131#issuecomment-114239926) we are still missing all the scripts, I personally find the current way (emake-based) more concise. But I will include the die calls, those totally make sense.

Did you write the ebuild yourself, or did you find it somewhere? If you're interested in helping out with neuroscience software on Gentoo, I'd sure need some help over here ^^

**Comment by TheChymera** [comment]
Regarding your latter question, ANTs is used to normalize brain imaging data (sort of how you would apply levels and lens deformation correction to a RAW image from your camera - only much more of a headache). You can find a short demo script (data included) [here](https://github.com/stnava/BasicBrainMapping).

**Comment by operatornormal** [comment]
Well, my version of ebuild install part was inspired by the way how debian packagers did on d/rules file but I'd say also that if software package  can install itself by "make install" it is the right and correct way ; this method of copying files from build tree is a kludge and prone to break if upstream modifies anything in the build process. 

I'm trying to get involved a little bit in gentoo maintenance too, so if there is anything to do in neuroscience-parts, I'd be glad to help. I know linux well, something about sw design and implementation (C, C++, java, c#, TCL, shell scripts) but next to nothing about neuroscience tools despite the fact my little sister used to operate a fMRI device :)

If there are well-defined tasks more related to computer science than neuroscience, I can start taking a look. ..maybe getting the makefile to copy the scripts too? My kludgish approach anyway produced an install that did include about same scripts that were included in debian install ; can't be that difficult from CMakefile.

I'm idling at irc using username costello but I do check github messages regularly too.

--
Antti

**Comment by TheChymera** [comment]
@operatornormal  hey again, and thanks for your offer of assistance!

I'm pretty much in the opposit situation - working in neuroscience for a while now, but pretty much banging my head against the keyboard where build systems are involved. Though recently I learnt to appreciate software that actually has any build system at all..

Anyway, perhaps the biggest issue currently on the neuroGentoo plate is writing an [ebuild for Freesurfer](https://github.com/gentoo-science/sci/issues/401). I am currently working on [its last major dependency](https://github.com/gentoo-science/sci/pull/442), and after that we could move on to freesurfer itself. But I guess you could also take a look beforehand and try to write an ebuild draft, if you are interested.

Regarding ANTs, I think we should be almost done. I just have to test the script installation again.

**Comment by operatornormal** [comment]
Aye, I'll clone the freesurfer tree to see how it behaves..

**Comment by TheChymera** [comment]
@jlec as far as I can tell this should be good to go!

**Comment by TheChymera** [comment]
@jlec bump?

**Comment by TheChymera** [comment]
@jlec @kiwifb ? could you help me figure out what the issue is with this package (if there still is one)?

**Comment by kiwifb** [comment]
I am sorry but I am somewhat overloaded and I keep taking stuff. I am about to open an issue about something closer to home that's adding to my load.

File: sci-biology/ants/ants-9999.ebuild [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+SLOT="0"
+LICENSE="BSD"
+KEYWORDS=""
+
+DEPEND="sci-libs/itk"
+RDEPEND="${DEPEND}"
+
+src_install() {
+	cd "${WORKDIR}/${P}_build/ANTS-build" || die "build dir not found"
+	emake DESTDIR="${D}" install
```
Comment (jlec) [inline sci-biology/ants/ants-9999.ebuild]: just call default here.

File: sci-biology/ants/ants-9999.ebuild [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	cd "${WORKDIR}/${P}/Scripts" || die "scripts dir not found"
+	dodir /usr/lib/ants
```
Comment (jlec) [inline sci-biology/ants/ants-9999.ebuild]: This isn't multilib safe. use ```get_libdir```here.

**Comment by TheChymera** [comment]
why, WHY does this always happen ?!?!?!?!

**Comment by TheChymera** [comment]
New pull request [here](https://github.com/gentoo-science/sci/pull/458).

**Pull request closed without merging**